File: week1/1/A.py
# https://leetcode.com/problems/group-anagrams/

class Solution:
    def groupAnagrams(strs: list[str]) -> list[list[str]]:
        # allocate dictionary 
        anagrams= {}

        # loop strs list
        for i in range(len(strs)):
            #  ex) sorted() return value ['a', 'a', 'd', 'g', 't', 'y']
            #  -> use join -> str type
            _str= ''.join(sorted(strs[i]))

            if _str not in anagrams:
                # add key: value in dictionary
                anagrams[_str]= [strs[i]]
            else:
                # append value in key "_str"
                anagrams[_str].append(strs[i])
        # from dictionary to list 
        return list(anagrams.values())


# 처음엔 아스키 코드의 합이 동일한 것을 이용해서 풀려고 시도했지만 인덱스 반환에서 실패
    def groupAnagrams2(strs: list[str]) -> list[list[str]]:
        _sum=0 
        sum_list= []
        for _str in strs:
            for c in _str:
                _sum+=ord(c)
            sum_list.append(_sum)
            _sum= 0

        print(sum_list)


result= Solution.groupAnagrams(["eat","tea","tan","ate","nat","bat"])
# result= Solution.groupAnagrams2(["eat","tea","tan","ate","nat","bat"])
print(result)

File: week1/1/A.rs
// From: A Po
// Author: Chris Chung
// Link: https://chungchris.github.io/2020/07/13/software/leetcode/Group-Anagrams/#toc-heading-4

use std::collections::HashMap;

pub fn group_anagrams(strs: Vec<String>) -> Vec<Vec<String>> {
    let base= 'a' as u32;
    let mut collect: HashMap<String, Vec<String>>= HashMap::new();
    for s in strs{
        let mut count: [i32; 26]= [0; 26];
        for c in s.chars(){
            count[(c as u32 - base) as usize ] += 1;
        }

        let id= format!("{:?}", count);
        match collect.get_mut(&id) {
            Some(r) => { r.push(s); }, // r is auto deref
            None => { collect.insert(id, vec![s]); }
        }
    }

    let mut ans:Vec<Vec<String>>= vec![];
    for (_ , v) in collect.drain(){
        ans.push(v);
    }

    ans
}

fn main(){
    let result= group_anagrams(vec!["eat".to_string(),"tea".to_string(),"tan".to_string()
        ,"ate".to_string(),"nat".to_string(),"bat".to_string()]);
    println!("{:?}", result)
}
